### PR TITLE
Add suse support (with mysql defaults this time)

### PR DESCRIPTION
**Pull request by ghormoon**
### PR progress checklist (to be filled in by reviewers)

- [ ] Changes to documentation are appropriate (or tick if not required)
- [ ] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

#### Primary type
- [x] `[feat]`     A new feature

### Does this PR introduce a `BREAKING CHANGE`?
No.

### Related issues and/or pull requests
Replaces https://github.com/saltstack-formulas/zabbix-formula/pull/140

### Describe the changes you're proposing
Following up on my colleagues work from https://github.com/saltstack-formulas/zabbix-formula/pull/140 i made adjustments to the defaults to go with mysql as you prefer. Please let me know if there are any other issues i did not address properly.

### CODE REVIEW

**Review by hatifnatt — APPROVED** [review]

**Review by myii — CHANGES_REQUESTED** [review]
Thanks for this PR, @ghormoon.  Unfortunately, this is failing with conflicting IDs in the CI.  Please uncomment these two lines in order to test on openSUSE during this PR's development:

https://github.com/saltstack-formulas/zabbix-formula/blob/25de37ebf0b6f0d272cd688c0c4a7ce617b67608/.gitlab-ci.yml#L145-L146

I've run these in my fork and you can see the results so far:

* https://gitlab.com/myii/zabbix-formula/-/jobs/1186061144
* https://gitlab.com/myii/zabbix-formula/-/jobs/1186061146

```sls
       local:
           Data failed to compile:
       ----------
           Detected conflicting IDs, SLS IDs need to be globally unique.
           The conflicting ID is 'zabbix_repo' and is found in SLS 'base:zabbix.agent.repo' and SLS 'base:zabbix.server.repo'
       ----------
           Detected conflicting IDs, SLS IDs need to be globally unique.
           The conflicting ID is 'zabbix_repo' and is found in SLS 'base:zabbix.server.repo' and SLS 'base:zabbix.frontend.repo'
```

File: zabbix/repo.sls
Comment (hatifnatt) [inline zabbix/repo.sls:57]: Missed this part, thanks @myii for pointing this issue. Variable must be used in state id to prevent duplicatees.
```suggestion
{{ id_prefix }}_repo:
```

**Review by hatifnatt — CHANGES_REQUESTED** [review]

File: zabbix/repo.sls
Comment (myii) [reply zabbix/repo.sls:57]: @hatifnatt Yes, unfortunately that's not enough, getting some strange errors after fixing that:

* https://gitlab.com/myii/zabbix-formula/-/jobs/1186085485
* https://gitlab.com/myii/zabbix-formula/-/jobs/1186085486

```sls
       [ERROR   ] An un-handled exception was caught by salt's global exception handler:
       StopIteration: 
       Traceback (most recent call last):
         File "/usr/bin/salt-call", line 8, in <module>
           sys.exit(salt_call())
         File "/usr/lib/python3.8/site-packages/salt/scripts.py", line 449, in salt_call
           client.run()
         File "/usr/lib/python3.8/site-packages/salt/cli/call.py", line 58, in run
           caller.run()
         File "/usr/lib/python3.8/site-packages/salt/cli/caller.py", line 112, in run
           ret = self.call()
         File "/usr/lib/python3.8/site-packages/salt/cli/caller.py", line 219, in call
           ret["return"] = self.minion.executors[fname](
         File "/usr/lib/python3.8/site-packages/salt/loader.py", line 1235, in __call__
           return self.loader.run(run_func, *args, **kwargs)
         File "/usr/lib/python3.8/site-packages/salt/loader.py", line 2268, in run
           return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
         File "/usr/lib/python3.8/site-packages/salt/loader.py", line 2283, in _run_as
           return _func_or_method(*args, **kwargs)
         File "/usr/lib/python3.8/site-packages/salt/executors/direct_call.py", line 12, in execute
           return func(*args, **kwargs)
         File "/usr/lib/python3.8/site-packages/salt/loader.py", line 1235, in __call__
           return self.loader.run(run_func, *args, **kwargs)
         File "/usr/lib/python3.8/site-packages/salt/loader.py", line 2268, in run
           return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
         File "/usr/lib/python3.8/site-packages/salt/loader.py", line 2283, in _run_as
           return _func_or_method(*args, **kwargs)
         File "/usr/lib/python3.8/site-packages/salt/modules/state.py", line 1115, in highstate
           ret = st_.call_highstate(
         File "/usr/lib/python3.8/site-packages/salt/state.py", line 4527, in call_highstate
           return self.state.call_high(high, orchestration_jid)
         File "/usr/lib/python3.8/site-packages/salt/state.py", line 3251, in call_high
           high, ext_errors = self.reconcile_extend(high)
         File "/usr/lib/python3.8/site-packages/salt/state.py", line 1627, in reconcile_extend
           state_type = next(x for x in body if not x.startswith("__"))
       StopIteration
       Traceback (most recent call last):
         File "/usr/bin/salt-call", line 8, in <module>
           sys.exit(salt_call())
         File "/usr/lib/python3.8/site-packages/salt/scripts.py", line 449, in salt_call
           client.run()
         File "/usr/lib/python3.8/site-packages/salt/cli/call.py", line 58, in run
           caller.run()
         File "/usr/lib/python3.8/site-packages/salt/cli/caller.py", line 112, in run
           ret = self.call()
         File "/usr/lib/python3.8/site-packages/salt/cli/caller.py", line 219, in call
           ret["return"] = self.minion.executors[fname](
         File "/usr/lib/python3.8/site-packages/salt/loader.py", line 1235, in __call__
           return self.loader.run(run_func, *args, **kwargs)
         File "/usr/lib/python3.8/site-packages/salt/loader.py", line 2268, in run
           return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
         File "/usr/lib/python3.8/site-packages/salt/loader.py", line 2283, in _run_as
           return _func_or_method(*args, **kwargs)
         File "/usr/lib/python3.8/site-packages/salt/executors/direct_call.py", line 12, in execute
           return func(*args, **kwargs)
         File "/usr/lib/python3.8/site-packages/salt/loader.py", line 1235, in __call__
           return self.loader.run(run_func, *args, **kwargs)
         File "/usr/lib/python3.8/site-packages/salt/loader.py", line 2268, in run
           return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
         File "/usr/lib/python3.8/site-packages/salt/loader.py", line 2283, in _run_as
           return _func_or_method(*args, **kwargs)
         File "/usr/lib/python3.8/site-packages/salt/modules/state.py", line 1115, in highstate
           ret = st_.call_highstate(
         File "/usr/lib/python3.8/site-packages/salt/state.py", line 4527, in call_highstate
           return self.state.call_high(high, orchestration_jid)
         File "/usr/lib/python3.8/site-packages/salt/state.py", line 3251, in call_high
           high, ext_errors = self.reconcile_extend(high)
         File "/usr/lib/python3.8/site-packages/salt/state.py", line 1627, in reconcile_extend
           state_type = next(x for x in body if not x.startswith("__"))
       StopIteration
```

Comment (myii) [reply zabbix/repo.sls:57]: @ghormoon @hatifnatt Looks like these `extend` blocks need to handle `Suse` as well?

* https://github.com/saltstack-formulas/zabbix-formula/search?q=extend+filename%3Arepo.sls

Comment (myii) [reply zabbix/repo.sls:57]: @ghormoon I see what the problem is in the `repo.sls` files (the `extends` haven't been set correctly).  Just trying something in my fork and will link inline with the results.

Comment (hatifnatt) [reply zabbix/repo.sls:57]: Yes there is a problem with extend
https://gitlab.com/myii/zabbix-formula/-/jobs/1186085485#L1687-#L1691 it's rendering as simple `{}` that's does not look like correct yaml.

File: zabbix/agent/repo.sls
Comment (myii) [inline zabbix/agent/repo.sls:26]:
```suggestion
```

This current method also splits the `RedHat` block, which will cause problems there.

All this file needs is the line above to be changed:

```diff
-    {% if salt['grains.get']('os_family') == 'Debian' -%}
+    {% if salt['grains.get']('os_family') in ['Debian', 'Suse'] -%}
```

File: zabbix/frontend/repo.sls
Comment (myii) [inline zabbix/frontend/repo.sls:26]: [Likewise](https://github.com/saltstack-formulas/zabbix-formula/pull/146/files#r614781594).

File: zabbix/proxy/repo.sls
Comment (myii) [inline zabbix/proxy/repo.sls:26]: [Likewise](https://github.com/saltstack-formulas/zabbix-formula/pull/146/files#r614781594).

File: zabbix/agent/repo.sls
Comment (myii) [reply zabbix/agent/repo.sls:26]: Note, this single line change is also required here:

https://github.com/saltstack-formulas/zabbix-formula/blob/c9aeb377d070cae54aa82f15904ab799b5994980/zabbix/server/repo.sls#L15

File: zabbix/repo.sls
Comment (myii) [reply zabbix/repo.sls:57]: @hatifnatt Actually, the current code isn't even hitting the `{}`, it's actually splitting the `RedHat` block with a `Suse` block.  I've mentioned the fixes below.

**Comment by myii** [comment]
@ghormoon After applying the fixes mentioned above, the good news is that this is _almost_ passing on openSUSE Leap but will need some more work on openSUSE Tumbleweed:

* https://gitlab.com/myii/zabbix-formula/-/jobs/1186140108
  - Just one test needs fixing.
* https://gitlab.com/myii/zabbix-formula/-/jobs/1186140109
  - Repo URL would need adjusting.

**Comment by hatifnatt** [comment]
>     * https://gitlab.com/myii/zabbix-formula/-/jobs/1186140109
>       
>       * Repo URL would need adjusting.

That's because Tumbleweed `grains['osmajorrelease']` != SLES `grains['osmajorrelease']` (at least I think so, can't verify).

@ghormoon are you using openSUSE Tumbleweed or SLES?

**Comment by ghormoon** [comment]
I'm using SLES 15, but honestly i didn't run the kitchen test, i just tried it manually :)
the problem is not sles vs opensuse, but the version number. if you look on the log, it shows 
https://repo.zabbix.com/zabbix/4.4/sles/20210414/x86_64/

my link is https://repo.zabbix.com/zabbix/5.2/sles/15/x86_64/ on my SLES 15 VM (the 5.2 is because of variable), so the problem is the tumbleweed versioning.
Either we can somehow hack it around if version>15 then version = 15, but the question is, if we want to run it on tumbleweed, as zabbix doesn't have repo for that. 99% the sles 15 repo will work, though.

leap is much closer to sles, so it makes sense to fix the path for files expected in tests 



i'll have a look on the suggestions and try to apply and test them locally on sles, i just got to it and you're quite active :)

**Comment by myii** [comment]
> That's because Tumbleweed `grains['osmajorrelease']` != SLES `grains['osmajorrelease']` (at least I think so, can't verify).

@hatifnatt @ghormoon I've tried a quick fix locally (using Kitchen) and there's deeper issues with Tumbleweed, for example:

```sls
       [INFO    ] Executing state pkg.installed for [zabbix-server]
       [DEBUG   ] Could not LazyLoad pkg.check_db: 'pkg.check_db' is not available.
       [DEBUG   ] targeting package: zabbix-server-mysql
       [DEBUG   ] Calling Zypper: systemd-run --scope zypper --non-interactive --no-refresh --no-refresh install --auto-agree-with-licenses --name zabbix-server-mysql
       [INFO    ] Executing command ['systemd-run', '--scope', 'zypper', '--non-interactive', '--no-refresh', '--no-refresh', 'install', '--auto-agree-with-licenses', '--name', 'zabbix-server-mysql'] in directory '/root'
       [ERROR   ] Command '['systemd-run', '--scope', 'zypper', '--non-interactive', '--no-refresh', '--no-refresh', 'install', '--auto-agree-with-licenses', '--name', 'zabbix-server-mysql']' failed with return code: 4
       [ERROR   ] stdout: Warning: The flag --no-refresh can only be used once.
       Loading repository data...
       Reading installed packages...
       Resolving package dependencies...
       
       Problem: nothing provides libevent-2.1.so.6()(64bit) needed by zabbix-server-mysql-4.4.10-1.el15.x86_64
        Solution 1: do not install zabbix-server-mysql-4.4.10-1.el15.x86_64
        Solution 2: break zabbix-server-mysql-4.4.10-1.el15.x86_64 by ignoring some of its dependencies
       
       Choose from above solutions by number or cancel [1/2/c/d/?] (c): c
       [ERROR   ] stderr: Running scope as unit: run-r0a507cfdfb0f48a7bca8fd3fdc509a86.scope
       [ERROR   ] retcode: 4
       [ERROR   ] An error was encountered while installing package(s): Zypper command failure: Running scope as unit: run-r0a507cfdfb0f48a7bca8fd3fdc509a86.scope
```

So let's not worry about Tumbleweed for now and just get Leap-15 up and running.  In fact, I'll just fix the test and push the fixes here and then this PR can be merged.

**Comment by myii** [comment]
Just for reference, this was the fix I was going to suggest (but not going to apply now):

```sls
{%- elif salt['grains.get']('os_family') == 'Suse' %}
{#-   TODO: When the v5 `map.jinja` is introduced to this formula, this should be provided via. the relevant `oscodename` parameter file #}
{%-   set majorrelease = '15' if grains.get('oscodename', '') == 'openSUSE Tumbleweed' else grains['osmajorrelease'] %}
{{ id_prefix }}_repo:
  pkgrepo.managed:
    - name: zabbix
    - humanname: "Zabbix Official Repository"
    - baseurl: https://repo.zabbix.com/zabbix/{{ zabbix.version_repo }}/sles/{{ majorrelease }}/x86_64/
    - gpgcheck: 1
    - gpgkey: https://repo.zabbix.com/zabbix/{{ zabbix.version_repo }}/sles/{{ majorrelease }}/x86_64/repodata/repomd.xml.key
    - gpgautoimport: True
```

**Comment by ghormoon** [comment]
yeah, it would be asking for dependency problems like you have in the log sooner or later. even if we were lucky now and it passed, we might get and bug reported in a month or two, so i'd stick to that only leap/sles is supported.
hacking it around like this is like using debian 10 packages on testing/experimental. it might work, but the further it diverges, the higher is the chance of issues

**Comment by myii** [comment]
@ghormoon OK, here's the final bit then, to get Leap-15 passing in GitLab CI:

https://github.com/saltstack-formulas/zabbix-formula/blob/2c67b11ce1b0ef4442a0d3dd1ec7bd3e7fb500ef/test/integration/default/controls/config_spec.rb#L41-L48

Please add the `suse` block as follows:

```rb
  case platform[:family]
  when 'debian'
    server_file_group = 'root'
    server_file_mode = '0644'
    setting_dbsocket = '/var/run/mysqld/mysqld.sock'
  when 'fedora'
    server_file_group = 'zabbixsrv'
  when 'suse'
    setting_dbsocket = '/run/mysql/mysql.sock'
  end
```

Then this PR should be ready for merging.

Would you prefer to do it or shall I push all of the changes myself?

**Comment by ghormoon** [comment]
ok, i tried to apply the changes and allowing leap on ci, we'll see if it passes now :)
thanks for the help so far

**Review by hatifnatt — APPROVED** [review]

**Review by myii — APPROVED** [review]

**Comment by myii** [comment]
Thanks, @ghormoon -- all working nicely.
@hatifnatt I went ahead and merged, since you already had got your approval in -- I wouldn't usually get involved when there's a codeowner available, so please don't think you have to wait for me!

**Pull request merged**
